includes/ProductSync/ProductValidator.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+namespace SkyVerge\WooCommerce\Facebook\ProductSync;
+
+use SkyVerge\WooCommerce\Facebook\Products;
+use WC_Product;
+use WC_Facebookcommerce_Integration;
+
+/**
+ * Class ProductValidator
+ *
+ * This class is responsible for validating whether a product should be synced to Facebook.
+ *
+ * @since 2.5.0
+ */
+class ProductValidator {
+
+	/**
+	 * @var string the meta key used to flag whether a product should be synced in Facebook
+	 */
+	const SYNC_ENABLED_META_KEY = '_wc_facebook_sync_enabled';
+
+	/**
+	 * @var WC_Facebookcommerce_Integration
+	 */
+	protected $integration;
+
+	/**
+	 * ProductValidator constructor.
+	 *
+	 * @param WC_Facebookcommerce_Integration $integration
+	 */
+	public function __construct( WC_Facebookcommerce_Integration $integration ) {
+		$this->integration = $integration;
+	}
+
+	/**
+	 * Validate whether a given product should be synced to Facebook.
+	 *
+	 * @param WC_Product $product A product object. Accepts both variations and variable products.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	public function validate( WC_Product $product ) {
+		if ( $this->integration->is_product_sync_enabled() ) {
+			throw new ProductExcludedException( 'Product sync is globally disabled.' );
+		}
+
+		$this->validate_product_status( $product );
+		$this->validate_product_stock_status( $product );
+		$this->validate_product_sync_field( $product );
+		$this->validate_product_price( $product );
+		$this->validate_product_visibility( $product );
+		$this->validate_product_categories_and_tags( $product );
+	}
+
+	/**
+	 * Check whether the product's status excludes it from sync.
+	 *
+	 * @param WC_Product $product A product object.
+	 *
+	 * @throws ProductExcludedException If product should not be synced.
+	 */
+	protected function validate_product_status( WC_Product $product ) {
+		$product = $product->get_parent_id() ? wc_get_product( $product->get_parent_id() ) : $product;
```

Comment:
We have multiple wc_get_product which is quite costly for our needs. It should be cached ( or not, we would need to verify ) but still, there is an overhead.

Reply:
Any database calls would be object cached but we could get the parent product in the class constructor to cut the product object instantiation from 4 down to 1.

Reply:
See f3cf0d7417bcc492597aef23d392ff4f6c076136 for a version that only gets the parent product once. I think this is cleaner too.